frontend/src/utils/websocket.js

```diff
@@ -1,611 +1,2886 @@
+      transport: this.constructor.name,
+      lastActivity: this.lastActivity,
+      pendingUpdates: Object.keys(this.pendingUpdates).length,
+      metrics: { ...this.metrics },
+      uptime: this.connectTime ? performance.now() - this.connectTime : 0,
+      messageEncoder: MessageEncoder.getStats(),
+      stateManager: this.stateManager.getMetrics(),
+      bulkOperations: {
+        totalBulkUpdates: this.metrics.bulkUpdatesProcessed,
+        totalBulkChanges: this.metrics.totalBulkChanges,
+        avgProcessingTime: this.metrics.avgBulkProcessingTime,
+        lastBulkSize: this.metrics.lastBulkProcessed,
+        lastBulkChanges: this.metrics.lastBulkChanged,
+        bulkEfficiency: this.metrics.lastBulkProcessed > 0 ?
+          (this.metrics.lastBulkChanged / this.metrics.lastBulkProcessed * 100).toFixed(1) + '%' : 'N/A'
+      }
+    };
+
+    // Add transport-specific optimization metrics
+    if (this.constructor.name === 'WebSocketClient') {
+      baseMetrics.transportOptimization = {
+        batchingEnabled: this.batchingEnabled,
+        queuedMessages: this.messageQueue.length,
+        maxQueueSize: this.maxQueueSize,
+        batchSize: this.batchSize,
+        batchDelay: this.batchDelay,
+        compression: {
+          messagesCompressed: this.compressionStats.messagesCompressed,
+          avgCompressionRatio: (this.compressionStats.avgCompressionRatio * 100).toFixed(1) + '%'
+        }
+      };
+    } else if (this.constructor.name === 'PostMessageClient') {
+      baseMetrics.transportOptimization = {
+        batchingEnabled: this.batchingEnabled,
+        queuedMessages: this.messageQueue.length,
+        maxQueueSize: this.maxQueueSize,
+        batchSize: this.batchSize,
+        batchDelay: this.batchDelay,
+        serialization: {
+          messagesOptimized: this.serializationStats.messagesOptimized,
+          avgSizeReduction: (this.serializationStats.avgSizeReduction * 100).toFixed(1) + '%'
+        }
+      };
     }
-    return this.worker.listFilesInDirectory(directoryPath);
+
+    return baseMetrics;
   }
 
-  // 2. run an arbitrary python script ------------
-  async runScript(scriptPath) {
-    console.log('[Client] runScript', scriptPath);
-    if (!this.isConnected || !this.worker) {
-      throw new Error('Connection not ready');
-    }
-    const result = await this.worker.runScript(scriptPath);
-    if (!result.success) {
-      throw new Error('Script execution failed');
-    }
-    this._handleComponentUpdate(result.components);
-    return result;
+  /**
+ * Enhanced error handling with MessageEncoder support
+ */
+  _handleError(error, context = '') {
+    this.metrics.errors++;
+    const errorMessage = `[${this.constructor.name}] ${context ? context + ': ' : ''}${error.message}`;
+    console.error(errorMessage, error);
+
+    // Create standardized error message
+    const errorPayload = {
+      type: 'error',
+      content: {
+        message: error.message,
```

Comment:
This is just a suggestion. Consider also include the stringified error for downstream inspection and logging.
```suggestion
message: error.message,
fullError: error.toString(),
```